### ISSUE TITLE
declare_dependency() should have a `variables` argument

### ISSUE THREAD

**Issue by dcbaker** [mesonbuild/meson#6047]
Since we allow variables with other types (cmake, pkg-config, config-tool), declare dependency should as well. I'm imagining that it would take a dictionary of key value pairs. Additionally the `get_variable` method would need to be extended with a keyword argument for internal dependencies.

The use case I'm wanting this for is subprojects, in particular, LLVM in mesa when used as a binary wrap.

Something like (subproject):
```meson
dep_to_export = delcare_depencency(
  dependencies : ...,
  include_directories : ...,
  variables : {
    resource_file : files('path/to/fresource),
    number_of_things : 2,
  }
)
```

consumer:
```meson
dep_foo = dependency('foo', fallback : ['foo', 'dep_to_export'])
myvar = dep_foo.get_variable(pkgconfig : 'resource_file_path', subproject : 'resource_file')
```

**Comment by dcbaker** [comment]
Before I start, @jpakkane, is this a feature you're interested in?

**Comment by jpakkane** [comment]
Yes, definitely. I thought there was already a bug on this. We definitely have talked about adding this.

It would be better, maintaining-wise, that the variable name should be the same for all backends. But I'm guessing existing projects already do different things with their pkgconfig/cmake setups?

**Comment by dcbaker** [comment]
Fixed by #6423
